Neos.RedirectHandler.DatabaseStorage/Classes/RedirectStorage.php

```diff
@@ -0,0 +1,247 @@
+<?php
+namespace Neos\RedirectHandler\DatabaseStorage;
+
+/*
+ * This file is part of the Neos.RedirectHandler.DatabaseStorage package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Doctrine\ORM\OptimisticLockException;
+use Neos\RedirectHandler\DatabaseStorage\Domain\Model\Redirect;
+use Neos\RedirectHandler\DatabaseStorage\Domain\Repository\RedirectRepository;
+use Neos\RedirectHandler\Exception;
+use Neos\RedirectHandler\Redirect as RedirectDto;
+use Neos\RedirectHandler\RedirectInterface;
+use Neos\RedirectHandler\Storage\RedirectStorageInterface;
+use Neos\RedirectHandler\Traits\RedirectSignalTrait;
+use TYPO3\Flow\Annotations as Flow;
+use TYPO3\Flow\Mvc\Routing\RouterCachingService;
+use TYPO3\Flow\Persistence\PersistenceManagerInterface;
+
+/**
+ * Database Storage for the Redirections
+ *
+ * @Flow\Scope("singleton")
+ */
+class RedirectStorage implements RedirectStorageInterface
+{
+    use RedirectSignalTrait;
+
+    /**
+     * @Flow\Inject
+     * @var RedirectRepository
+     */
+    protected $redirectRepository;
+
+    /**
+     * @Flow\Inject
+     * @var PersistenceManagerInterface
+     */
+    protected $persistenceManager;
+
+    /**
+     * @Flow\Inject
+     * @var RouterCachingService
+     */
+    protected $routerCachingService;
+
+    /**
+     * Runtime cache to avoid creating multiple time the same redirection
```

Comment:
Yes per example during a batch import, a CSV can contains duplicated redirect (with the same URI), this cache avoid key contraint errors